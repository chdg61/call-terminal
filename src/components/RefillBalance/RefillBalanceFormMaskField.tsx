import classnames from 'classnames';
import * as React from 'react';
import MaskedInput from 'react-text-mask';

interface Props {
    name: string;
    label: string;
    mask: any;
    value: string;
    error: string;
    touched: boolean;
    onChange?: React.ChangeEventHandler<HTMLInputElement>;
    onBlur?: React.FocusEventHandler<HTMLInputElement>;
    placeholder?: string;
}


export class RefillBalanceFormMaskField extends React.Component<Props> {

    constructor(props: Readonly<Props>) {
        super(props);

        this.onBlur = this.onBlur.bind(this);
        this.onChange = this.onChange.bind(this);
    }

    static defaultProps: Partial<Props> = {
        placeholder: '',
    }

    private onChange(e: React.ChangeEvent<HTMLInputElement>){
        if(this.props.onChange && typeof this.props.onChange === 'function'){
            this.props.onChange(e);
        }
    }

    private onBlur(e: React.FocusEvent<HTMLInputElement>){
        if(this.props.onBlur && typeof this.props.onBlur === 'function'){
            this.props.onBlur(e);
        }
    }

    render(): React.ReactElement<any, string | React.JSXElementConstructor<any>> | string | number | {} | React.ReactNodeArray | React.ReactPortal | boolean | null | undefined {
        const {
            name,
            label,
            mask,
            touched,
            error,
            value,
            placeholder
        } = this.props;

        return (
            <React.Fragment>
                <label htmlFor={name}>
                    {label}
                </label>
                <MaskedInput
                    mask={mask}
                    className={classnames('form-control', {
                        'is-invalid': error && touched,
                    })}
                    value={value}
                    id={name}
                    placeholder={placeholder}
                    onBlur={this.onBlur}
                    onChange={this.onChange}
                    name={name}
                    guide
                />
                {touched && !!error && (
                    <div className="invalid-feedback">
                        {error}
                    </div>
                )}
            </React.Fragment>
        );
    }
}
