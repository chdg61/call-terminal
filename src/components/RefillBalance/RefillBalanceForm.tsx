import { FormikBag, FormikProps, withFormik } from 'formik';
import * as React from 'react';
import createNumberMask from 'text-mask-addons/dist/createNumberMask';
import * as yup from 'yup';
import { RefillBalanceFormMaskField } from './RefillBalanceFormMaskField';

interface Props {
    onSubmit: (values: FormValues) => void;
}

export interface FormValues {
    phone: string;
    amount: string | number;
}

type AllProps = Props & FormikProps<FormValues>;

export class RefillBalanceForm extends React.Component<AllProps> {

    render(): React.ReactElement<any, string | React.JSXElementConstructor<any>> | string | number | {} | React.ReactNodeArray | React.ReactPortal | boolean | null | undefined {
        const {
            handleSubmit,
            handleChange,
            handleBlur,
            touched,
            errors,
            values
        } = this.props;

        return (
            <form onSubmit={handleSubmit}>
                <div className="form-group">
                    <RefillBalanceFormMaskField
                        mask={['+', '7', ' ', '(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/]}
                        label="Номер телефона"
                        name="phone"
                        placeholder="Введите номер телефона"
                        value={values.phone}
                        onBlur={handleBlur}
                        onChange={handleChange}
                        error={errors.phone}
                        touched={touched.phone}
                    />
                </div>
                <div className="form-group">
                    <RefillBalanceFormMaskField
                        label="Сумма"
                        mask={createNumberMask({
                            prefix: '',
                            suffix: ' ₽',
                            thousandsSeparatorSymbol: ' ',
                        })}
                        value={String(values.amount)}
                        error={errors.amount}
                        touched={touched.amount}
                        placeholder="Введите сумму для пополнения"
                        name="amount"
                        onBlur={handleBlur}
                        onChange={handleChange}
                    />
                </div>
                <button type="submit" className="btn btn-primary">
                    Пополнить
                </button>
            </form>
        );
    }
}

function transformAmount (amountStr = ''): string | number {
    return parseFloat(amountStr.replace(/\s/g, '')) || 0;
}

export const RefillBalanceFormWrapper = withFormik<Props, FormValues>({
    validationSchema: (props: Props) => {
        return yup.object().shape({
            phone: yup.string()
                .matches(
                    /^\+7\s\([1-9]\d{2}\)\s\d{3}-\d{4}/i,
                    'Телефон должен быть формата +7(912) 345-5566'
                )
                .required('Поле обязательное для заполнения'),
            amount: yup
                .number()
                .transform(function(value, originalvalue) {
                    if (typeof originalvalue === 'string') {
                        return transformAmount(originalvalue);
                    }

                    return 0;
                })
                .min(1, 'Минимальная сумма платежа - 1 рубль')
                .max(1000, 'Максимальная сумма платежа - 1000 рублей')
                .required('Поле обязательное для заполнения'),
        });
    },

    mapPropsToValues(props: Props): FormValues {
        return {
            amount: '',
            phone: '',
        };
    },

    handleSubmit(values: FormValues, formikBag:  FormikBag<Props, FormValues>) {
        formikBag.props.onSubmit({
            ...values,
            amount: transformAmount(String(values.amount)),
        });
    }
})(RefillBalanceForm);
