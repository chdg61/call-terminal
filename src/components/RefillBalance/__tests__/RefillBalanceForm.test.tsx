import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { RefillBalanceForm, RefillBalanceFormWrapper } from '../RefillBalanceForm';
import { RefillBalanceFormMaskField } from '../RefillBalanceFormMaskField';


describe('RefillBalanceFormWrapper', () => {

    it('empty form', () => {
        const onSubmit = jest.fn();
        const wrapper = shallow(
            <RefillBalanceFormWrapper onSubmit={onSubmit} />
        );
        const instance = wrapper.instance();

        expect(wrapper.html()).toMatchSnapshot();
    });

    describe('validate phone', () => {
        it('validate change phone number - bad value', (done) => {
            const onSubmit = jest.fn();
            const wrapper = mount(
                <RefillBalanceFormWrapper onSubmit={onSubmit} />
            );
            const formField = wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(0);
            formField.instance().onChange({
                persist: () => {},
                target: {
                    name: 'phone',
                    value: '+7 (123) 55_-____'
                }
            });
            formField.instance().onBlur({
                persist: () => {},
                target: {
                    name: 'phone',
                    value: '+7 (123) 55_-____'
                }
            });
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find(RefillBalanceForm).props().errors.phone).toBe('Телефон должен быть формата +7(912) 345-5566');
                expect(wrapper.find(RefillBalanceForm).props().touched.phone).toBeTruthy();
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(0).find('.invalid-feedback').length).toBe(1);
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(0).find('.invalid-feedback').text()).toBe('Телефон должен быть формата +7(912) 345-5566');
                done();
            })
        });

        it('validate change phone number - good value', (done) => {
            const onSubmit = jest.fn();
            const wrapper = mount(
                <RefillBalanceFormWrapper onSubmit={onSubmit} />
            );

            const formField = wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(0);
            formField.instance().onChange({
                persist: () => {},
                target: {
                    name: 'phone',
                    value: '+7 (123) 556-7788'
                }
            });
            formField.instance().onBlur({
                persist: () => {},
                target: {
                    name: 'phone',
                    value: '+7 (123) 556-7788'
                }
            });
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find(RefillBalanceForm).props().errors.phone).toBeUndefined();
                expect(wrapper.find(RefillBalanceForm).props().touched.phone).toBeTruthy();
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(0).find('.invalid-feedback').length).toBe(0);
                done();
            })
        });

        it('empty phone number', (done) => {
            const onSubmit = jest.fn();
            const wrapper = mount(
                <RefillBalanceFormWrapper onSubmit={onSubmit} />
            );
            const formField = wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(0);
            formField.instance().onBlur({
                persist: () => {},
                target: {
                    name: 'phone',
                    value: ''
                }
            });
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find(RefillBalanceForm).props().errors.phone).toBe('Поле обязательное для заполнения');
                expect(wrapper.find(RefillBalanceForm).props().touched.phone).toBeTruthy();
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(0).find('.invalid-feedback').length).toBe(1);
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(0).find('.invalid-feedback').text()).toBe('Поле обязательное для заполнения');
                done();
            })
        });
    });

    describe('validate amount', () => {
        it('validate change amount - min amount', (done) => {
            const onSubmit = jest.fn();
            const wrapper = mount(
                <RefillBalanceFormWrapper onSubmit={onSubmit} />
            );
            const formField = wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1);
            formField.instance().onChange({
                persist: () => {},
                target: {
                    name: 'amount',
                    value: '0 Р'
                }
            });
            formField.instance().onBlur({
                persist: () => {},
                target: {
                    name: 'amount',
                    value: '0 Р'
                }
            });
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find(RefillBalanceForm).props().errors.amount).toBe('Минимальная сумма платежа - 1 рубль');
                expect(wrapper.find(RefillBalanceForm).props().touched.amount).toBeTruthy();
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1).find('.invalid-feedback').length).toBe(1);
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1).find('.invalid-feedback').text()).toBe('Минимальная сумма платежа - 1 рубль');
                done();
            })
        });

        it('validate change amount - max amount', (done) => {
            const onSubmit = jest.fn();
            const wrapper = mount(
                <RefillBalanceFormWrapper onSubmit={onSubmit} />
            );
            const formField = wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1);
            formField.instance().onChange({
                persist: () => {},
                target: {
                    name: 'amount',
                    value: '1001 Р'
                }
            });
            formField.instance().onBlur({
                persist: () => {},
                target: {
                    name: 'amount',
                    value: '1001 Р'
                }
            });
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find(RefillBalanceForm).props().errors.amount).toBe('Максимальная сумма платежа - 1000 рублей');
                expect(wrapper.find(RefillBalanceForm).props().touched.amount).toBeTruthy();
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1).find('.invalid-feedback').length).toBe(1);
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1).find('.invalid-feedback').text()).toBe('Максимальная сумма платежа - 1000 рублей');
                done();
            })
        });

        it('validate change amount - good value', (done) => {
            const onSubmit = jest.fn();
            const wrapper = mount(
                <RefillBalanceFormWrapper onSubmit={onSubmit} />
            );

            const formField = wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1);
            formField.instance().onChange({
                persist: () => {},
                target: {
                    name: 'amount',
                    value: '1000 Р'
                }
            });
            formField.instance().onBlur({
                persist: () => {},
                target: {
                    name: 'amount',
                    value: '1000 Р'
                }
            });
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find(RefillBalanceForm).props().errors.amount).toBeUndefined();
                expect(wrapper.find(RefillBalanceForm).props().touched.amount).toBeTruthy();
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1).find('.invalid-feedback').length).toBe(0);
                done();
            })
        });

        it('empty amount', (done) => {
            const onSubmit = jest.fn();
            const wrapper = mount(
                <RefillBalanceFormWrapper onSubmit={onSubmit} />
            );
            const formField = wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1);
            formField.instance().onBlur({
                persist: () => {},
                target: {
                    name: 'amount',
                    value: ''
                }
            });
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find(RefillBalanceForm).props().errors.amount).toBe('Поле обязательное для заполнения');
                expect(wrapper.find(RefillBalanceForm).props().touched.amount).toBeTruthy();
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1).find('.invalid-feedback').length).toBe(1);
                expect(wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1).find('.invalid-feedback').text()).toBe('Поле обязательное для заполнения');
                done();
            })
        });
    });

    describe('submit form', () => {
        it('not submit because not valid phone or amount', (done) => {
            const onSubmit = jest.fn();
            const wrapper = mount(
                <RefillBalanceFormWrapper onSubmit={onSubmit} />
            );
            const form = wrapper.find(RefillBalanceForm).find('form');
            form.simulate("submit", );
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find(RefillBalanceForm).props().isValid).toBeFalsy();
                expect(onSubmit).not.toBeCalled();
                done();
            })
        });

        it('submit with valid phone and amount', (done) => {
            const onSubmit = jest.fn();
            const wrapper = mount(
                <RefillBalanceFormWrapper onSubmit={onSubmit} />
            );

            const phoneField = wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(0);
            phoneField.instance().onChange({
                persist: () => {},
                target: {
                    name: 'phone',
                    value: '+7 (123) 556-7788'
                }
            });
            phoneField.instance().onBlur({
                persist: () => {},
                target: {
                    name: 'phone',
                    value: '+7 (123) 556-7788'
                }
            });


            const amountField = wrapper.find(RefillBalanceForm).find(RefillBalanceFormMaskField).at(1);
            amountField.instance().onChange({
                persist: () => {},
                target: {
                    name: 'amount',
                    value: '1000 Р'
                }
            });
            amountField.instance().onBlur({
                persist: () => {},
                target: {
                    name: 'amount',
                    value: '1000 Р'
                }
            });

            const form = wrapper.find(RefillBalanceForm).find('form');
            form.simulate("submit", );
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find(RefillBalanceForm).props().isValid).toBeTruthy();
                expect(onSubmit).toBeCalled();
                expect(onSubmit).toBeCalledWith({
                    amount: 1000,
                    phone: '+7 (123) 556-7788'
                });
                done();
            })
        });
    })


});
