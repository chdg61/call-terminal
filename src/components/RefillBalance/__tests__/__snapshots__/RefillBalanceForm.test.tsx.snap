// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefillBalanceFormWrapper empty form 1`] = `"<form><div class=\\"form-group\\"><label for=\\"phone\\">Номер телефона</label><input class=\\"form-control\\" id=\\"phone\\" placeholder=\\"Введите номер телефона\\" name=\\"phone\\" value=\\"\\"/></div><div class=\\"form-group\\"><label for=\\"amount\\">Сумма</label><input class=\\"form-control\\" id=\\"amount\\" placeholder=\\"Введите сумму для пополнения\\" name=\\"amount\\" value=\\"\\"/></div><button type=\\"submit\\" class=\\"btn btn-primary\\">Пополнить</button></form>"`;
